tensorflow/stream_executor/rocm/rocm_dnn.cc

```diff
@@ -195,6 +215,124 @@ MIOPEN_DNN_ROUTINE_EACH(PERFTOOLS_GPUTOOLS_MIOPEN_WRAP)
+
+  const int kMaxMIOpenTensorSize = 5;
+
+  uint64 GetHashValue(miopenTensorDescriptor_t tensor_desc) {
+
+    miopenDataType_t dataType = miopenFloat;
+    int dims[kMaxMIOpenTensorSize] = {0};
+    int strides[kMaxMIOpenTensorSize] = {0};
+    miopenGetTensorDescriptor(tensor_desc, &dataType, dims, strides);
+
+    uint64 hashValue = tensorflow::hash<int>()(dataType);
+    for (int dim : dims) hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(dim));
+    for (int stride : strides) hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(stride));
+
+    return hashValue;
+  }
+
+  uint64 GetHashValue(miopenConvolutionDescriptor_t conv_desc) {
+
+    miopenConvolutionMode_t c_mode = miopenConvolution;
+    int pad_h = 0, pad_w = 0, u = 0, v = 0, dilation_h = 0, dilation_w = 0;
+    miopenGetConvolutionDescriptor(conv_desc, &c_mode, &pad_h, &pad_w, &u, &v, &dilation_h, &dilation_w);
+
+    uint64 hashValue = tensorflow::hash<int>()(c_mode);
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(pad_h));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(pad_w));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(u));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(v));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(dilation_h));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<int>()(dilation_w));
+
+    return hashValue;
+  }
+
+  uint64 GetHashValue(miopenActivationDescriptor_t actv_desc) {
+
+    miopenActivationMode_t mode = miopenActivationPASTHRU;
+    double alpha = 0.0, beta = 0.0, gamma = 0.0;
+    miopenGetActivationDescriptor(actv_desc, &mode, &alpha, &beta, &gamma);
+
+    uint64 hashValue = tensorflow::hash<int>()(mode);
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<double>()(alpha));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<double>()(beta));
+    hashValue = tensorflow::Hash64Combine(hashValue, tensorflow::hash<double>()(gamma));
+
+    return hashValue;
+  }
+
+  // class to implement a cache of compiled fusion plans
+  class CachedFusionPlans {
+
+  public:
+
+    // check if we already have a fusion_plan corresponding to the given hash value
+    // If we fo
+    //   return true (+ the cached fusion plan via given pointer)
+    // Else
+    //   create a new fusion plan descriptor,
+    //   associate it with the given hash value in the cache
+    //   return false (+ newly created fusion plan via given pointer)
+    static bool FindOrCreate(uint64 hash, ROCMExecutor* parent, miopenFusionPlanDescriptor_t *fusion_plan,
+			     miopenFusionDirection_t fusion_direction, miopenTensorDescriptor_t input_descriptor) {
+
+      mutex_lock lock{cachedPlansMutex};
+
+      bool foundCachedPlan = false;
+
+      auto it = cachedPlans.find(hash);
+      if (it != cachedPlans.end()) {
```

Comment:
could you re-indent the code with `git clang-format`?

Reply:
will be part of next commit...
just ran it...now I realize why all the TF code does not contain long lines...have to say not a fan of clang-format, makes code difficult too read!